unc_classification_tab_data/models/vae.py

```diff
@@ -0,0 +1,325 @@
+
+        Parameters
+        ----------
+        input_tensor: torch.Tensor
+            The input to the encoder.
+        """
+        h = self.hidden(input_tensor)
+        mean = self.mean(h)
+        log_std = self.log_std(h)
+        std = self.std(log_std)
+        return mean, std
+
+
+class Decoder(nn.Module):
+    """The decoder module, which decodes a sample from the latent space back to the space of
+    the input data.
+
+    Parameters
+    ----------
+    input_dim: int
+        The dimensionality of the input
+    hidden_dims: List[int]
+        A list with the sizes of the hidden layers.
+    z_dim: int
+        The size of the latent space.
+    """
+
+    def __init__(self, input_dim: int, hidden_dims: List[int], z_dim: int):
+        super().__init__()
+        architecture = [z_dim] + hidden_dims
+        self.hidden_layers = []
+        if hidden_dims == []:
+            self.hidden_layers = []
+        else:
+            for i in range(len(architecture) - 1):
+                self.hidden_layers.append(nn.Linear(architecture[i], architecture[i + 1]))
+                self.hidden_layers.append(nn.ReLU())
+        self.hidden = nn.Sequential(*self.hidden_layers)
+        self.mean = nn.Linear(architecture[-1], input_dim)
+        self.log_std = nn.Linear(architecture[-1], input_dim)
+        self.std = nn.Sigmoid()
+
+    def forward(self, input_tensor: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Perform forward pass of decoder. Returns mean and standard deviation corresponding
+        to an independent Normal distribution.
+
+        Parameters
+        ----------
+        input_tensor: torch.Tensor
+            A sample from the latent space, which has to be decoded.
+        """
+        h = self.hidden(input_tensor)
+        mean = self.mean(h)
+        log_std = self.log_std(h)
+        std = torch.sqrt(torch.exp(log_std))
+        return mean, std
+
+
+class VAEModule(nn.Module):
+    """The Pytorch module of a Variational Autoencoder, consisting of an equally-sized encoder and
+    decoder. This module works for continuous distributions. In case of discrete distributions,
+    it has to be adjusted (outputting a Bernoulli distribution instead of independent Normal).
+
+    Parameters
+    ----------
+    input_dim: int
+        The dimensionality of the input, assumed to be a 1-d vector.
+    hidden_dims: List[int]
+        A list of integers, representing the hidden dimensions of the encoder and decoder. These
+        hidden dimensions are the same for the encoder and the decoder.
+    z_dim: int
+        The dimensionality of the latent space.
+    """
+
+    def __init__(self, input_dim: int, hidden_dims: List[int], z_dim: int):
+        super().__init__()
+
+        self.z_dim = z_dim
+        self.encoder = Encoder(input_dim, hidden_dims, z_dim)
+        self.decoder = Decoder(input_dim, hidden_dims, z_dim)
+
+    def forward(self, input_tensor: torch.Tensor, reconstr_error_weight) -> Tuple[
```

Comment:
missing typing for weight, also for readability:

```
def forward(
    self, input_tensor: torch.Tensor, reconstr_error_weight: float
) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
```

Reply:
done